File: ai/linear_model.R
#!/usr/bin/env Rscript
#
# Command-line linear regression analysis.
#
# Usage:
#   Rscript linear_model.R <filename> <x_column> <y_column>
#
# Example:
#   Rscript linear_model.R regression_data.csv YearsExperience Salary

suppressPackageStartupMessages({
  library(ggplot2)
})

# Parse and validate command-line arguments.
parse_arguments <- function() {
  args <- commandArgs(trailingOnly = TRUE)

  if (length(args) != 3) {
    cat(
      "Usage: Rscript linear_model.R <filename> <x_column> <y_column>\n",
      file = stderr()
    )
    quit(status = 1)
  }

  list(
    filename = args[1],
    x_column = args[2],
    y_column = args[3]
  )
}

# Read CSV and validate that requested columns exist.
load_data <- function(filename, x_column, y_column) {
  if (!file.exists(filename)) {
    cat(sprintf("Error: file not found: %s\n", filename), file = stderr())
    quit(status = 1)
  }

  data <- tryCatch(
    read.csv(filename, stringsAsFactors = FALSE),
    error = function(e) {
      cat(sprintf("Error reading file: %s\n", e$message), file = stderr())
      quit(status = 1)
    }
  )

  missing <- setdiff(c(x_column, y_column), names(data))
  if (length(missing) > 0) {
    cat(
      sprintf("Error: column(s) not found: %s\n", paste(missing, collapse = ", ")),
      file = stderr()
    )
    cat(
      sprintf("Available columns: %s\n", paste(names(data), collapse = ", ")),
      file = stderr()
    )
    quit(status = 1)
  }

  data
}

# Fit linear regression and compute evaluation metrics.
fit_regression <- function(data, x_column, y_column) {
  formula <- as.formula(paste(y_column, "~", x_column))
  model <- lm(formula, data = data)

  slope <- coef(model)[x_column]
  intercept <- coef(model)["(Intercept)"]
  predictions <- fitted(model)

  x_values <- data[[x_column]]
  y_values <- data[[y_column]]

  r_value <- cor(x_values, y_values, method = "pearson")
  mse <- mean((y_values - predictions)^2)
  r_squared <- summary(model)$r.squared

  list(
    slope = slope,
    intercept = intercept,
    r_value = r_value,
    mse = mse,
    r_squared = r_squared,
    predictions = predictions,
    x_values = x_values,
    y_values = y_values
  )
}

# Print regression statistics to standard output.
print_statistics <- function(slope, intercept, r_value, mse, r_squared) {
  cat("Linear Regression Results\n")
  cat(paste(rep("=", 40), collapse = ""), "\n")
  cat(sprintf("Slope:                 %.4f\n", slope))
  cat(sprintf("Intercept:             %.4f\n", intercept))
  cat(sprintf("Pearson r:             %.4f\n", r_value))
  cat(sprintf("Mean Squared Error:    %.4f\n", mse))
  cat(sprintf("R-squared:             %.4f\n", r_squared))
  cat(sprintf("Equation:              y = %.4f x + %.4f\n", slope, intercept))
}

# Create scatter plot with regression line and annotations.
create_plot <- function(
  data,
  x_column,
  y_column,
  slope,
  intercept,
  r_value,
  output_path
) {
  plot_data <- data.frame(
    x = data[[x_column]],
    y = data[[y_column]]
  )

  equation_text <- sprintf("y = %.2f x + %.2f", slope, intercept)
  correlation_text <- sprintf("r = %.4f", r_value)
  annotation_text <- paste(equation_text, correlation_text, sep = "\n")

  p <- ggplot(plot_data, aes(x = x, y = y)) +
    geom_point(alpha = 0.7, color = "steelblue", size = 3) +
    geom_smooth(method = "lm", se = FALSE, color = "red", linewidth = 1) +
    annotate(
      "label",
      x = min(plot_data$x) + 0.05 * diff(range(plot_data$x)),
      y = max(plot_data$y),
      label = annotation_text,
      hjust = 0,
      vjust = 1,
      fill = "white",
      alpha = 0.85,
      size = 4
    ) +
    labs(
      title = paste(y_column, "vs", x_column),
      x = x_column,
      y = y_column
    ) +
    theme_minimal()

  ggsave(output_path, plot = p, width = 10, height = 6, dpi = 150)
}

main <- function() {
  args <- parse_arguments()
  data <- load_data(args$filename, args$x_column, args$y_column)

  results <- fit_regression(data, args$x_column, args$y_column)
  print_statistics(
    results$slope,
    results$intercept,
    results$r_value,
    results$mse,
    results$r_squared
  )

  create_plot(
    data,
    args$x_column,
    args$y_column,
    results$slope,
    results$intercept,
    results$r_value,
    "regression_plot_r.png"
  )

  cat("\nPlot saved to regression_plot_r.png\n")
}

main()

File: ai/linear_model.py
#!/usr/bin/env python3
"""
Command-line linear regression analysis.

Usage:
    python linear_model.py <filename> <x_column> <y_column>

Example:
    python linear_model.py regression_data.csv YearsExperience Salary
"""

import sys

import matplotlib.pyplot as plt
import pandas as pd
from scipy.stats import pearsonr
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


def parse_arguments(argv):
    """Validate and return command-line arguments."""
    if len(argv) != 4:
        print(
            "Usage: python linear_model.py <filename> <x_column> <y_column>",
            file=sys.stderr,
        )
        sys.exit(1)
    return argv[1], argv[2], argv[3]


def load_data(filename, x_column, y_column):
    """Read CSV and validate that requested columns exist."""
    try:
        data = pd.read_csv(filename)
    except FileNotFoundError:
        print(f"Error: file not found: {filename}", file=sys.stderr)
        sys.exit(1)
    except pd.errors.EmptyDataError:
        print(f"Error: file is empty: {filename}", file=sys.stderr)
        sys.exit(1)

    missing = [col for col in (x_column, y_column) if col not in data.columns]
    if missing:
        print(
            f"Error: column(s) not found: {', '.join(missing)}",
            file=sys.stderr,
        )
        print(f"Available columns: {', '.join(data.columns)}", file=sys.stderr)
        sys.exit(1)

    return data


def fit_regression(x, y):
    """Fit linear regression and compute evaluation metrics."""
    x_values = x.values.reshape(-1, 1)
    y_values = y.values

    model = LinearRegression()
    model.fit(x_values, y_values)
    predictions = model.predict(x_values)

    slope = model.coef_[0]
    intercept = model.intercept_
    r_value, _ = pearsonr(x_values.ravel(), y_values)
    mse = mean_squared_error(y_values, predictions)
    r_squared = r2_score(y_values, predictions)

    return slope, intercept, r_value, mse, r_squared, predictions


def print_statistics(slope, intercept, r_value, mse, r_squared):
    """Print regression statistics to standard output."""
    print("Linear Regression Results")
    print("=" * 40)
    print(f"Slope:                 {slope:.4f}")
    print(f"Intercept:             {intercept:.4f}")
    print(f"Pearson r:             {r_value:.4f}")
    print(f"Mean Squared Error:    {mse:.4f}")
    print(f"R-squared:             {r_squared:.4f}")
    print(f"Equation:              y = {slope:.4f}x + {intercept:.4f}")


def create_plot(
    x,
    y,
    predictions,
    slope,
    intercept,
    r_value,
    x_column,
    y_column,
    output_path,
):
    """Create scatter plot with regression line and annotations."""
    fig, ax = plt.subplots(figsize=(10, 6))

    ax.scatter(x, y, alpha=0.7, edgecolors="black", linewidth=0.5, label="Observations")
    ax.plot(x, predictions, color="red", linewidth=2, label="Regression line")

    equation_text = f"y = {slope:.2f}x + {intercept:.2f}"
    correlation_text = f"r = {r_value:.4f}"

    ax.text(
        0.05,
        0.95,
        f"{equation_text}\n{correlation_text}",
        transform=ax.transAxes,
        fontsize=12,
        verticalalignment="top",
        bbox=dict(boxstyle="round", facecolor="white", alpha=0.85),
    )

    ax.set_xlabel(x_column)
    ax.set_ylabel(y_column)
    ax.set_title(f"{y_column} vs {x_column}")
    ax.legend()
    ax.grid(True, linestyle="--", alpha=0.4)

    fig.tight_layout()
    fig.savefig(output_path, dpi=150)
    plt.close(fig)


def main():
    filename, x_column, y_column = parse_arguments(sys.argv)
    data = load_data(filename, x_column, y_column)

    x = data[x_column]
    y = data[y_column]

    slope, intercept, r_value, mse, r_squared, predictions = fit_regression(x, y)
    print_statistics(slope, intercept, r_value, mse, r_squared)
    create_plot(
        x,
        y,
        predictions,
        slope,
        intercept,
        r_value,
        x_column,
        y_column,
        "regression_plot_python.png",
    )
    print("\nPlot saved to regression_plot_python.png")


if __name__ == "__main__":
    main()
